ui: Statements tracing - Demo

koorosh opened the pull request:
Current PR is not supposed to be merged. For demo purposes only.

koorosh committed:
ui: (fix) Apply styles for navigation panes

Before, Text component styles were adjusted to style
links on side navigation panel. It was wrong because
Text component should not care about particular styling
of navigation pane.
Now, particular styles which relate to navigation panel
only is moved to SideNavigation component.
It is changed now because current implementation breaks
styles for Diagnostics panel.

Release note: None

koorosh committed:
ui: Link component

Link component is a wrapper on top of <a> element
is created to maintain consistent styles for links.

Release note: None

koorosh committed:
ui: Add Diagnostics view for Statements

- Adjust SummaryCard to provide external styles
- Add DiagnosticsView component which renders
view when no diagnostics available.
- Current implementation includes only view without
any logic connected to redux store.

Release note: None

koorosh committed:
ui: Add Diagnostics view for Statements

- Adjust SummaryCard to provide external styles
- Add DiagnosticsView component which renders
view when no diagnostics available.
- Current implementation includes only view without
any logic connected to redux store.

Release note: None

koorosh committed:
ui: Add Diagnostics view for Statements

- Adjust SummaryCard to provide external styles
- Add DiagnosticsView component which renders
view when no diagnostics available.
- Current implementation includes only view without
any logic connected to redux store.

Release note: None

koorosh committed:
ui: Add sagas to handle statement tracing actions

Statements sagas are implemented to be able request
statement diagnostics and retrieve when its ready.

This commit defines Actions for requesting diagnostics
and sagas which listens for those actions.

Also selectors are added to find statement by
statement id (query) which contains diagnostic data.

Release note: None

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Add tooltips for diagnostics statuses

- All code related to statements diagnostics are moved to
statements/diagnostics dir;
- Status badges are refactored to external component which is
responsible for displaying proper text of status and text for
tooltips.
- Tooltips have to be shown on some views and disabled on other.
To implement this, `enableTooltip` property is added. Particulary
it is needed to hide tooltip for "READY" status in Diagnostics tab
and show it on Statements List view.

Release note (admin ui): Show tooltips for statement diagnostic statuses

koorosh committed:
ui: Add tooltips for diagnostics statuses

- All code related to statements diagnostics are moved to
statements/diagnostics dir;
- Status badges are refactored to external component which is
responsible for displaying proper text of status and text for
tooltips.
- Tooltips have to be shown on some views and disabled on other.
To implement this, `enableTooltip` property is added. Particulary
it is needed to hide tooltip for "READY" status in Diagnostics tab
and show it on Statements List view.

Release note (admin ui): Show tooltips for statement diagnostic statuses

koorosh committed:
ui: Add tooltips for diagnostics statuses

- All code related to statements diagnostics are moved to
statements/diagnostics dir;
- Status badges are refactored to external component which is
responsible for displaying proper text of status and text for
tooltips.
- Tooltips have to be shown on some views and disabled on other.
To implement this, `enableTooltip` property is added. Particulary
it is needed to hide tooltip for "READY" status in Diagnostics tab
and show it on Statements List view.

Release note (admin ui change): Show tooltips for statement diagnostic statuses

koorosh committed:
ui: Create HeaderSection shared component

It is common for pages to have header with 'back' navigation link.
HeaderSection component wraps page header with logic of navigation
back to specified page via props.

Release note: None

koorosh committed:
ui: Extend Statements table with Diagnostics column

Current changes add column which doesn't consume any
data and always shows 'Activate' link.
It is starting point for next changes:
- providing diagnostics statuses;
- displaying modal for requesting diagnostics;

Release note (admin ui): Add Diagnostics column to Statements Page

koorosh committed:
ui: Add Modal component

- Modal component is a wrapper around Ant.design Modal component
with customized styles;

Release note: None

koorosh committed:
ui: Integrate ActivateDiagnosticsModal with Statements page

ActivateDiagnosticsModal is implemented with total encapsulation
of modal's state and connected props to store into component.
- ActivateDiagnosticsModal component instance exposes `showModalFor`
method through forwardedRef that makes possible to manage open/closed
state of modal inside of component and parent component can trigger
where modal should be opened and provide required arguments.
Current pattern is more anti-pattern but with this particular case
it make easy to separate responsibilities between components.
- ActivateDiagnosticsModal component is connected to store and
dispatches action to request new diagnostics;

Statements table is based on SortedTable component which doesn't
provide easy way to handle clicks on rows (due to some magic
bootstrapping process). This required to propagate method which
triggers modal opening through several functions outside of
component.

Release note (admin ui change): Open Activate Diagnostics modal from Statements page (Activate link)

koorosh committed:
ui: Rename Link component to Anchor

Release note: None

koorosh committed:
ui: Integrate requestStatementDiagnostics API

Replace faked API with real implementation to
request statement diagnostics.

Release note: None

koorosh committed:
ui: Add selectDiagnosticRequestsByStatementFingerprint selector

Release note: None

koorosh committed:
ui: Connect Statement Diagnostics API

Before, Statements diagnostics view used faked API
to simulate responses from server.
Now redux API has changed to comply with actual
implementation of API.

Release note: None

koorosh committed:
ui: Connect Statement Diagnostics API

Before, Statements diagnostics view used faked API
to simulate responses from server.
Now redux API has changed to comply with actual
implementation of API.

Release note: None

koorosh committed:
ui: Connect Statement Diagnostics API wit diagnostics table component

Replacing faked API responses with real implementation required some
redesigns with following important changes:

- Requesting diagnosticsRequests initiated when DiagnosticsView is opened.
Then to request new diagnostics requests before existing data is invalidated,
we need to manually invalidate and then refresh data.

- Current backend API doesn't support Error state for requested data so only
"Ready", and "Waiting for query" statuses are possible.

Release note: None

koorosh committed:
ui: Add DownloadFile component

DownloadFile component accepts file name and content and
then initiates file download from client.
It works as a wrapper to catch onClick event on wrapped
component.

Release note: None

koorosh committed:
ui: Refactor Statement diagnostics action ()redux) names

Release note: None

koorosh committed:
ui: Add traces downloading on Statement Diagnostics tab

- Add ability to download json file with tracing data
- Refactor code to comply with changed backend API
- add `long` package because exposed interface for
getting statement diagnostics require diagnostics id
as Long type. Current library provides compatible
implementation of Long type with protobuf.js interfaces.

Release note: None

koorosh committed:
ui: Add Statement Diagnostics History page

- Page contains list of all available statement diagnostics reports
- Added link on Debug page to reference on Statement Diagnostics History page
- Added link on Statement Diagnostics tab to reference on Statement Diagnostics History page

Release note (admin ui change): Statement Diagnostics History page added

koorosh committed:
ui: Add links on Statement diagnostics documentation

Release note: None

koorosh committed:
ui: Remove View Trace button as functionality is not ready

Release note: None

koorosh committed:
ui: Apply sorting for Statement Diagnostics table and Diagnostics History page

- Sort by 'Activated On' column by default

Release note (admin ui change): Statement Diagnostics table, default sorting

koorosh committed:
ui: AutoClose AlertMessage on timeout

Before, alert messages could be closed only manually.
With current implementation, Alert component allows
to specify if it can be closed with timeout (6 seconds),
and if it can be closed manually (hide close button on alert).

Release note: None

koorosh committed:
ui: Show alert message when user requests new diagnostics report

- Show success alert message on Statements and Diagnostics View pages
when request is successful. Success message will be closed automatically
after 6 sec.
- In case of error - show failed message and requires to be closed manually.
- All messages will be closed when user leaves page.

Release note: None

koorosh committed:
ui: Fit Diagnostics column in Statements table

Statements table columns were styled to occupy all
width and adding one more column stretched table
width beyond the screen. To fit it, styles for columns
with bar charts were changed (decreased min width)

Release note: None

koorosh committed:
ui: Rename tracing file to statement-diagnostics.json

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45800)
<!-- Reviewable:end -->

koorosh committed:
ui: Add handlers for downloading file from Statement Diagnostics History page

Logic for downloading file with traces was missed
on Diagnostics History page. It is the same functionality
as on Statement Diagnostics tab.

Release note: None

koorosh committed:
Squashed commit of the following:

commit b486a6a62ff407dfca893e9821d6c0b490e4b845
Author: Andrii Vorobiov <and.vorobiov@gmail.com>
Date:   Thu Feb 27 11:52:38 2020 +0200

    ui: Extend Badge component styles

    Add styles for Badge statuses: success, danger, and info.
    These styles are defined in Design system.

    Release note: None

Release note (<category, see below>): <what> <show> <why>

koorosh committed:
Squashed commit of the following:

commit 73120cd3ea7df94989f3bb51bff109bc06a79c8f
Author: Andrii Vorobiov <and.vorobiov@gmail.com>
Date:   Thu Feb 27 11:58:12 2020 +0200

    ui: Extend Button component styles

    - Add styles for button secondary type (according to DS)
    - Add styles to prevent word wrapping inside button. Button will
    have width to place its content in one line
    - Fix styles for aligning button icon (left-right sides)

    Release note: None

Release note (<category, see below>): <what> <show> <why>

koorosh committed:
Squashed commit of the following:

commit b0ab7def584ff8e6e1a22eb9129e4b72dcb4de0d
Author: Andrii Vorobiov <and.vorobiov@gmail.com>
Date:   Mon Mar 2 15:38:08 2020 +0200

    ui: Extend Table component with pagination size props

    Allow passing pageSize prop to Table component to
    define how many row should be displayed per page.

    Release note: None

Release note (<category, see below>): <what> <show> <why>

koorosh committed:
Squashed commit of the following:

commit c4733bff62c39186a92f0112d9134f665e4f1df4
Author: Andrii Vorobiov <and.vorobiov@gmail.com>
Date:   Mon Mar 2 15:29:50 2020 +0200

    ui: Extend Text component with 'code' type

    Add styles to support monospace text by Text component.

    Release note: None

Release note (<category, see below>): <what> <show> <why>

koorosh committed:
Squashed commit of the following:

commit feb3d45b8235fb53f426606916da4db3ec1f4051
Author: David Hartunian <davidh@cockroachlabs.com>
Date:   Mon Mar 2 22:53:52 2020 -0500

    sql: add API for accessing statement diagnostics

    Added 3 new RPC endpoints under `_status` for manipulating Statement Diagnostics
    Requests:

    1. `CreateStatementDiagnosticsRequest`
    Creates a new request for a given statement fingerprint

    2. `StatementDiagnosticsRequests`
    Fetches all statement diagnostics requests

    3. `StatementDiagnostics`
    Fetches a specific statement diagnostics request that has been completed and
    includes the JSON payload of the trace

    Basic implementations for these endpoints have been added to the
    `apiReducers.tsx` and `api.ts` files for use by the Admin UI.

    A minor bug where completed requests did not have `completedAt` timestamp set
    was fixed.

commit 9d684ea68c6c28b40192be552d7d3009eeef8321
Merge: 2e621c7c81 62c8667569
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Wed Mar 4 02:09:24 2020 +0000

    Merge #45575

    45575: cli: reveal SQL error details/hints in CLI commands r=ajwerner a=knz

    Found while investigating #43114. First commit from that PR.

    Previously, when a CLI command that uses SQL under the hood
    encountered an error, only the main error message was displayed.

    This patch changes it to reveal the other additional user-visible
    fields.

    Release note (cli change): Certain CLI command now provide more
    details and/or a hint when they encounter an error.

    Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>

commit 2e621c7c81b077adb295d7487fe15b32eeb79a80
Merge: 7b64b0bf0e 3daeced65a
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Wed Mar 4 00:26:33 2020 +0000

    Merge #45520

    45520: opt: FK checks for Upsert r=RaduBerinde a=RaduBerinde

    These commit add support for opt-driven FK checks for UPSERT and INSERT ON CONFLICT DO UPDATE. This is a simplified and cleaned up reimplementation of #45095.

    There aren't a lot of logictests around FKs and upsert. I plan to try to build a randomized testing facility for FK checks, where we generate randomized mutations and cross-check FK violations or lack there of against a separate instance of the database without FK relationships.

    #### Revert "opt: add first half of upsert FK checks"

    This reverts commit 96447c8d2e11ad888d8c39b990c33fea18f4ed9f.

    We are reconsidering the direction of the upsert FK checks: instead of filtering
    inserted vs updated rows, we can use the return columns to get the "new" values.
    This change also made some things harder to understand: we used to build the FK
    check input using just the columns corresponding to the FK instead of all
    columns.

    Release note: None

    #### opt: refactor optbuilder FK check code

    The optbuilder FK check code has become unnecessarily complicated:
     - there are two ways of creating a WithScan (`makeFKInputScan` and
       `projectOrdinals`) which are similar but take different kinds of information
       as input.
     - there are various slices of column ordinals or column IDs that are threaded
       through long parts of code, making it hard to follow.

    This change cleans this up by creating a fkCheckHelper which contains the
    metadata related to FK table ordinals and is capable of creating a WithScan with
    either the "new" or the "fetched" values.

    The new code should generate the same effective plans; the differences are:
     - we now always create the WithScan before the other table Scan so some column IDs in the plan changed;
     - we no longer have an unnecessary projection for Update (it was used to
       renumber the columns coming out of WithScan).

    Release note: None

    #### opt: add outbound FK checks for upsert

    This is a re-implementation of Justin's change 3d1dd0fdfb, except that we now
    perform the insertion check for all new rows instead of just the inserted rows.
    We do this by utilizing the same column that would be used for a RETURNING
    clause, which in some cases is of the form
    `CASE WHEN canary IS NULL col1 ELSE col2`.

    Release note: None

    #### opt: add inbound FK checks for upsert

    This change adds inbound FK checks for upsert and switches execution over to the
    new style FK checks for upsert.

    Similar to UPDATE, the inbound FK checks run on the set difference between "old"
    values for the FK columns and "new" values.

    Release note (performance improvement): improved execution plans of foreign key
    checks for UPSERT and INSERT ON CONFLICT in some cases (in particular
    multi-region).

    Co-authored-by: Radu Berinde <radu@cockroachlabs.com>

commit 3daeced65acf43287826955c4a707463a13e4200
Author: Radu Berinde <radu@cockroachlabs.com>
Date:   Tue Mar 3 16:23:50 2020 -0800

    opt: add inbound FK checks for upsert

    This change adds inbound FK checks for upsert and switches execution over to the
    new style FK checks for upsert.

    Similar to UPDATE, the inbound FK checks run on the set difference between "old"
    values for the FK columns and "new" values.

    Release note (performance improvement): improved execution plans of foreign key
    checks for UPSERT and INSERT ON CONFLICT in some cases (in particular
    multi-region).

commit b8ebc3ac1e3d846fcc8e3cae3377e8744ddc04d4
Author: Radu Berinde <radu@cockroachlabs.com>
Date:   Tue Mar 3 16:23:50 2020 -0800

    opt: add outbound FK checks for upsert

    This is a re-implementation of Justin's change 3d1dd0fdfb, except that we now
    perform the insertion check for all new rows instead of just the inserted rows.
    We do this by utilizing the same column that would be used for a RETURNING
    clause, which in some cases is of the form
    `CASE WHEN canary IS NULL col1 ELSE col2`.

    Release note: None

commit 75e6009938970111055506c870ac3fdce655fae8
Author: Radu Berinde <radu@cockroachlabs.com>
Date:   Tue Mar 3 16:23:50 2020 -0800

    opt: refactor optbuilder FK check code

    The optbuilder FK check code has become unnecessarily complicated:
     - there are two ways of creating a WithScan (`makeFKInputScan` and
       `projectOrdinals`) which are similar but take different kinds of information
       as input.
     - there are various slices of column ordinals or column IDs that are threaded
       through long parts of code, making it hard to follow.

    This change cleans this up by creating a fkCheckHelper which contains the
    metadata related to FK table ordinals and is capable of creating a WithScan with
    either the "new" or the "fetched" values.

    The new code should generate the same effective plans; the differences are:
     - we now always create the WithScan before the other table Scan so some column IDs in the plan changed;
     - we no longer have an unnecessary projection for Update (it was used to
       renumber the columns coming out of WithScan).

    Release note: None

commit aa2295dcbd069b0644a9bc717228316adf689505
Author: Radu Berinde <radu@cockroachlabs.com>
Date:   Tue Mar 3 16:23:50 2020 -0800

    Revert "opt: add first half of upsert FK checks"

    This reverts commit 96447c8d2e11ad888d8c39b990c33fea18f4ed9f.

    We are reconsidering the direction of the upsert FK checks: instead of filtering
    inserted vs updated rows, we can use the return columns to get the "new" values.
    This change also made some things harder to understand: we used to build the FK
    check input using just the columns corresponding to the FK instead of all
    columns.

    Release note: None

commit 7b64b0bf0e443b3d30c5aa7652961ee2c924cbb0
Merge: ede97bc8b3 5b1598d95b
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 23:33:09 2020 +0000

    Merge #45513

    45513: sql: disable primary key changes when a schema change is in progress r=lucy-zhang a=rohany

    Fixes #45362.

    Release note (sql change): This PR disables primary key changes
    when a concurrent schema change is executing on the same table,
    or if a schema change has been started on the same table in
    the current transaction.

    Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>

commit ede97bc8b36b24c2f04e78af56eeae68ea23b1b8
Merge: bbbb26bd9d 7383e9dcd3
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 22:44:11 2020 +0000

    Merge #45597

    45597: opt: extend WithUses and improve NeededMutationCols r=RaduBerinde a=RaduBerinde

    This change extends the WithUses rule prop to keep track of the columns actually
    used by `WithScan`s.

    This set is then used by NeededMutationCols to make sure we never prune mutation
    input columns that are used by FK checks. Currently this never happens, but
    because of fairly subtle reasons: FKs happen to require indexes on both sides,
    and those indexes force the mutation operator to require values for those
    columns. This won't be the case anymore with upsert FK checks, for which this
    fix is required.

    The new information will also be used (in a future change) to prune unused
    columns of `With` itself.

    Release note: None

    Co-authored-by: Radu Berinde <radu@cockroachlabs.com>

commit bbbb26bd9d3b9e70f76455011a3b9b386c11c362
Merge: e1e1f2cd7c d80411f4ff
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 21:58:08 2020 +0000

    Merge #45397

    45397: sql: disallow other schema changes during an in progress primary key change r=lucy-zhang a=rohany

    Fixes #45363.

    Release note (sql change): This commit disables the use of schema
    changes when a primary key change is in progress. This includes
    the following:
    * running a primary key change in a transaction, and then starting
      another schema change in the same transaction.
    * starting a primary key change on one connection, and then starting
      a schema change on the same table on another connection while
      the initial primary key change is currently executing.

    Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>

commit e1e1f2cd7c4b892421ba159e6bf4fd27200e4ff4
Merge: 064c4ea34b f38087bf11 66d502f641 2b7aa4691b 35095d9192
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 20:36:07 2020 +0000

    Merge #45472 #45607 #45609 #45613

    45472: storage/engine: Teeing engine fixes r=itsbilal a=itsbilal

    This change introduces some misc teeing engine fixes, such as proper
    cache initialization, copying of SSTs before ingestions (now that
    pebble also deletes SSTs after an Ingest), and better null msg
    handling in GetProto. The cache part fixes a segfault in
    GetStats.

    Fixes #42654 .

    Release note: None.

    45607: sql: fix usages of ActiveVersionOrEmpty in {create, alter}_table.go r=irfansharif a=rohany

    Fixes #45519.

    This PR removes an incorrect usage of ActiveVersionOrEmpty in
    alter_table.go, and updates a comment in create_table.go detailing its
    usage.

    Release note: None

    45609: sql: support star-expanding label-less tuples + numeric tuple indexing r=RaduBerinde a=knz

    Requested by @RaduBerinde  / @ajwerner .

    Previously, it was impossible to use `(t).*` to expand a single
    tuple-typed column into multiple projections if the tuple was not
    labeled to start with.

    This limitation was caused by another limitation, that it was not
    possible to refer to a single column in a tuple if the tuple was not
    already labeled.

    This patch addresses both limitations.

    Release note (sql change): CockroachDB now supports expanding all
    columns of a tuple using the `.*` notation, for example `SELECT (t).*
    FROM (SELECT (1,'b',2.3) AS t)`. This is a CockroachDB-specific extension.

    Release note (sql change): CockroachDB now supports accessing the Nth
    column in a column with tuple type using the syntax `(...).@N`, for
    example `SELECT (t).@2 FROM (SELECT (1,'b',2.3) AS t)`. This is a
    CockroachDB-specific extension.

    45613: pgwire: fix decimal decoding with trailing zeroes r=jordanlewis a=otan

    Resolves https://github.com/cockroachdb/cockroach/issues/25460

    In addition to the release note, I have also modified the docker runfile
    to support elixir tests, and also updated the elixir tests to be run.

    Release note (bug fix): In previous PRs, drivers that do not truncate
    trailing zeroes for decimals in the binary format end up having
    inaccuracies of up to 10^4 during the decode step. In this PR, we fix
    the error by truncating the trailing zeroes as appropriate. This fixes
    known incorrect decoding cases with Postgrex in Elixir.

    Co-authored-by: Bilal Akhtar <bilal@cockroachlabs.com>
    Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>
    Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>
    Co-authored-by: Oliver Tan <otan@cockroachlabs.com>

commit 35095d919220724f97039d5921318d032a74661e
Author: Oliver Tan <otan@cockroachlabs.com>
Date:   Mon Mar 2 15:40:45 2020 -0800

    pgwire: fix decimal decoding with trailing zeroes

    In addition to the release note, I have also modified the docker runfile
    to support elixir tests, and also updated the elixir tests to be run.

    Release note (bug fix): In previous PRs, drivers that do not truncate
    trailing zeroes for decimals in the binary format end up having
    inaccuracies of up to 10^4 during the decode step. In this PR, we fix
    the error by truncating the trailing zeroes as appropriate. This fixes
    known incorrect decoding cases with Postgrex in Elixir.

commit 064c4ea34b33b26894cbf1fde3cd31135e247455
Merge: ff2b6055b1 665312778c
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 19:08:46 2020 +0000

    Merge #45630

    45630: storageccl: rework TestRandomKeyAndTimestampExport to be shorter r=pbardea a=ajwerner

    The test was taking a very long time due to the tiny page sizes. This commit
    changes the test to scale the total number of keys based on the page size.

    Before:

    ```
    --- PASS: TestRandomKeyAndTimestampExport (19.06s)
    ```

    After:
    ```
    --- PASS: TestRandomKeyAndTimestampExport (2.30s)
    ```

    Release note: None

    Co-authored-by: Andrew Werner <ajwerner@cockroachlabs.com>

commit f38087bf11391d4525b250f53d0d29cb1abdac2a
Author: Bilal Akhtar <bilal@cockroachlabs.com>
Date:   Wed Feb 26 15:45:56 2020 -0500

    storage/engine: Teeing engine fixes

    This change introduces some misc teeing engine fixes, such as proper
    cache initialization, copying of SSTs before ingestions (now that
    pebble also deletes SSTs after an Ingest), and better null msg
    handling in GetProto. The cache part fixes a segfault in
    GetStats.

    It also unifies file handling between in-memory and on-disk
    teeing engines, by ensuring we write to files in each engine's
    aux directory if we're writing to one engine's aux directory.

    Fixes #42654 .

    Release note: None.

commit 5b1598d95bfddb6771bc62bf05e96e3998a5bb06
Author: Rohan Yadav <rohany@alumni.cmu.edu>
Date:   Thu Feb 27 12:42:24 2020 -0500

    sql: disable primary key changes when a schema change is in progress

    Fixes #45362.

    Release note (sql change): This PR disables primary key changes
    when a concurrent schema change is executing on the same table,
    or if a schema change has been started on the same table in
    the current transaction.

commit d80411f4ff9c03615aa3ddd2b238d8d4a255ba39
Author: Rohan Yadav <rohany@alumni.cmu.edu>
Date:   Tue Feb 25 10:59:17 2020 -0500

    sql: disallow other schema changes an in progress primary key change

    Fixes #45363.

    Release note (sql change): This commit disables the use of schema
    changes when a primary key change is in progress. This includes
    the following:
    * running a primary key change in a transaction, and then starting
      another schema change in the same transaction.
    * starting a primary key change on one connection, and then starting
      a schema change on the same table on another connection while
      the initial primary key change is currently executing.

commit ff2b6055b114929fb424bfe7948243d1be126d70
Merge: aeb41dc981 9dcab01f6c cee9355126
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 16:59:33 2020 +0000

    Merge #45347 #45502

    45347: sql: make secondary indexes not write empty k/v's + bugfixes for primary key changes r=jordanlewis a=rohany

    Fixes #45223.

    Depends on #45226 to land first.

    This PR fixes many bugs around secondary index encodings
    and CRUD operations k/v reads and writes.

    * Fixes a problem secondary indexes would write empty
      k/v's if it contained a family that had all null values.
    * Fix multiple bugs where updates to a table during an online
      primary key change could result an inconsistent
      new primary key.
    * Fix an assumption in the updater that assumed indexes
      always had the same number of k/v's. The logic has been
      updated to perform a sort of merge operation to decide
      what k/v's to insert/delete during the update operation.
    * Increased testing around secondary indexes k/vs and
      schema change operations.

    The release note is None because these are all bugs
    introduced in 20.1.

    Release note: None

    45502: sql: allow rename database for sequences without a db name reference r=rohany a=otan

    Resolves immediate concern from https://github.com/cockroachdb/cockroach/issues/45411
    Refs: https://github.com/cockroachdb/cockroach/issues/34416

    See release note for description. This PR should be included ahead of
    the more "general" fix of changing the DefaultExpr with the new database
    as it unblocks people using
    `experimental_serial_normalization=sql_sequence` from using the database
    rename feature.

    Release note (sql change): Previously, when we renamed a database, any
    table referencing a sequence would be blocked from being able to rename
    the table. This is to block cases where if the table's reference to the
    sequence contains the database name, and the database name changes, we
    have no way of overwriting the table's reference to the sequence in the
    new database. However, if no database name is included in the sequence
    reference, we should continue to allow the database to rename, as is
    implemented with this change.

    Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>
    Co-authored-by: Oliver Tan <otan@cockroachlabs.com>

commit 665312778c3b705f7d2b4a4a8d75a1d761c166d1
Author: Andrew Werner <ajwerner@cockroachlabs.com>
Date:   Tue Mar 3 00:33:17 2020 -0500

    storageccl: rework TestRandomKeyAndTimestampExport to be shorter

    The test was taking a very long time due to the tiny page sizes. This commit
    changes the test to scale the total number of keys based on the page size.

    Before:

    ```
    --- PASS: TestRandomKeyAndTimestampExport (19.06s)
    ```

    After:
    ```
    --- PASS: TestRandomKeyAndTimestampExport (2.30s)
    ```

    Release note: None

commit 66d502f641b2cc133bc3e73d89685d989d374082
Author: Rohan Yadav <rohany@alumni.cmu.edu>
Date:   Mon Mar 2 18:01:59 2020 -0500

    sql: fix usages of ActiveVersionOrEmpty in {create, alter}_table.go

    Fixes #45519.

    This PR removes an incorrect usage of ActiveVersionOrEmpty in
    alter_table.go, and updates a comment in create_table.go detailing its
    usage.

    Release note: None

commit aeb41dc98115da5d160e1cc01d2fa9c9e9c14c01
Merge: ada086ebde 3d7aeb361d 0a1f2516c2
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 16:10:07 2020 +0000

    Merge #45595 #45603

    45595: sql: make cleanup jobs spawned by alter primary key not cancelable r=spaskob a=rohany

    Fixes #45500.

    This PR makes the job spawned by ALTER PRIMARY KEY that cleans
    up indexes be uncancelable.

    This PR additionally fixes a related bug where ALTER PRIMARY KEY
    would spawn a job when it didn't actually enqueue any mutations
    on the table descriptor, causing future schema changes on the
    table to hang.

    Release note (sql change): This PR makes the cleanup job spawned
    by ALTER PRIMARY KEY in some cases uncancelable.

    45603: storage/txnwait: terminate push when pusher aborted at lower epoch r=nvanbenschoten a=nvanbenschoten

    Closes #40786.
    Closes #44336.

    This commit resolves a bug in distributed deadlock detection that would
    allow a deadlock between transactions to go undetected, stalling the
    workload indefinitely.

    The issue materialized as follows:
    1. two transactions would deadlock and each enter a txnwait queue
    2. they would poll their pushees record along with their own
    3. deadlock detection would eventually pick this up and abort one of the txns
       using the pusher's copy of the txn's proto
    4. however, the aborted txn has since restarted and bumped it epoch
    5. the aborted txn continued to query its record, but failed to ingest any
       updates from it because the record was at a lower epoch than its own
       copy of its txn proto. So it never noticed that it was ABORTED
    6. all other txns in the system including the original contending txn
       piled up behind the aborted txn in the contention queue, waiting for
       it to notice it was aborted and exit the queue
    7. deadlock!

    I'm optimistically closing the two `kv/contention/nodes=4` issues both
    because I hope this is the cause of their recent troubles and also because
    I've been spending a lot of time with the test recently in light of #45482
    and plan to stabilize it fully.

    I plan to backport this to release-19.2. This doesn't need to go all the
    way back to release-19.1 because this was introduces in aed892a.

    Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>
    Co-authored-by: Nathan VanBenschoten <nvanbenschoten@gmail.com>

commit ada086ebde92b35de12d66823d23faea85808606
Merge: b0be21ac55 80894c3fc3 4caee85de6 7e0ba7c697
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 15:26:58 2020 +0000

    Merge #44130 #45589 #45604

    44130: sql: postgresql dollar-quoted string support r=knz a=damienhollis

    Added support for postgresql dollar-quoted strings in the scanner. A
    dollar-quoted string constant consists of a dollar sign ($), an
    optional "tag" of zero or more characters, another dollar sign, an
    arbitrary sequence of characters that makes up the string content, a
    dollar sign, the same tag that began this dollar quote, and a final
    dollar sign.

    The scanner uses the existing token type of SCONST for dollar-quoted
    strings. As a result, when the AST is formatted as a string, there is
    no knowledge that the original input was dollar-quoted and it is
    therefore formatted as either a plain string or an escaped
    string (depending on the content).

    Fixes #41777.

    Release Note (sql change): CockroachDB now supports string and byte
    array literals using the dollar-quoted notation, as documented here:
    https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-DOLLAR-QUOTING

    45589: sql: update error message for primary key change on an interleave parent r=otan a=rohany

    Fixes #45537.

    This PR updates the error message when trying to perform a primary
    key change on an interleaved parent to include the name of the
    table as well as the names of the interleaved children.

    Release note: None

    45604: opt: factor limit hints into scan and lookup join costs r=rytaft a=rytaft

    This PR is a continuation of #43415. The first commit is copied directly
    from that PR, and the second commit includes a minor fix as well as a
    number of test cases.

    Fixes #34811; the example query in this issue now chooses a lookup join
    as desired. The coster now takes limit hints into account when costing
    scans and lookup joins, and propagates limit hints through lookup joins.

    Release note (sql change): The optimizer now considers the likely number
    of rows an operator will need to provide, and might choose query plans
    based on this. In particular, the optimizer might prefer lookup joins
    over alternatives in some situations where all rows of the join will
    probably not be needed.

    Co-authored-by: damien.hollis <damien.hollis@unimarket.com>
    Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>
    Co-authored-by: Céline O'Neil <celineloneil@gmail.com>
    Co-authored-by: Rebecca Taft <becca@cockroachlabs.com>

commit 7e0ba7c6972fb5dd91f1216139606b3b40195a96
Author: Rebecca Taft <becca@cockroachlabs.com>
Date:   Mon Mar 2 15:35:36 2020 -0600

    opt: make lookup join limit hint consistent with coster, add tests

    This commit adds a new helper function called lookupJoinInputLimitHint, which
    is called by both the coster and the physical props code for limit hint
    calculation. This helper function ensures that both places take into account
    the batch size of 100, and require that the calculated limit hint is a
    multiple of this batch size.

    This commit also adds more tests related to costing of limit hints for
    scans and lookup joins.

    Release note: None

commit 9dcab01f6c1f1fe8a35a43e486f5ebccee178a36
Author: Rohan Yadav <rohany@alumni.cmu.edu>
Date:   Thu Feb 20 15:37:29 2020 -0500

    sql: make secondary indexes not write empty k/v's + bugfixes for primary
    key changes

    Fixes #45223.

    Depends on #45226 to land first.

    This PR fixes many bugs around secondary index encodings
    and CRUD operations k/v reads and writes.

    * Fixes a problem secondary indexes would write empty
      k/v's if it contained a family that had all null values.
    * Fix multiple bugs where updates to a table during an online
      primary key change could result an inconsistent
      new primary key.
    * Fix an assumption in the updater that assumed indexes
      always had the same number of k/v's. The logic has been
      updated to perform a sort of merge operation to decide
      what k/v's to insert/delete during the update operation.
    * Increased testing around secondary indexes k/vs and
      schema change operations.

    The release note is None because these are all bugs
    introduced in 20.1.

    Release note: None

commit 80894c3fc3dc7409601c53f93201646054a6d5ff
Author: damien.hollis <damien.hollis@unimarket.com>
Date:   Sun Jan 19 21:16:13 2020 +1300

    sql: support postgresql dollar-quoted strings

    Added support for postgresql dollar-quoted strings in the scanner. A
    dollar-quoted string constant consists of a dollar sign ($), an
    optional "tag" of zero or more characters, another dollar sign, an
    arbitrary sequence of characters that makes up the string content, a
    dollar sign, the same tag that began this dollar quote, and a final
    dollar sign.

    The scanner uses the existing token type of SCONST for dollar-quoted
    strings. As a result, when the AST is formatted as a string, there is
    no knowledge that the original input was dollar-quoted and it is
    therefore formatted as either a plain string or an escaped
    string (depending on the content).

    Fixes #41777.

    Release Note (sql change): CockroachDB now supports string and byte
    array literals using the dollar-quoted notation, as documented here:
    https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-DOLLAR-QUOTING

commit 3d7aeb361dcb9fe295e7685968c606ecb32a7ff0
Author: Rohan Yadav <rohany@alumni.cmu.edu>
Date:   Mon Mar 2 13:46:35 2020 -0500

    sql: make cleanup jobs spawned by alter primary key not cancelable

    Fixes #45500.

    This PR makes the job spawned by ALTER PRIMARY KEY that cleans
    up indexes be uncancelable.

    This PR additionally fixes a related bug where ALTER PRIMARY KEY
    would spawn a job when it didn't actually enqueue any mutations
    on the table descriptor, causing future schema changes on the
    table to hang.

    Release note (sql change): This PR makes the cleanup job spawned
    by ALTER PRIMARY KEY in some cases uncancelable.

commit b0be21ac55f839e98597a600f9f8479b0eccc5a4
Merge: 8ffade8456 bbac108247
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 14:48:15 2020 +0000

    Merge #45621

    45621: sql: allow inverted indexes on mixed-case cols r=jordanlewis a=jordanlewis

    Closes #42944.

    Previously, a bug prevented creation of inverted indexes on columns with
    mixed-case identifiers. Now, the bug is fixed.

    Release note (bug fix): it is now possible to create inverted indexes on
    columns whose names are mixed-case.

    Co-authored-by: Jordan Lewis <jordanthelewis@gmail.com>

commit 2b7aa4691bc7969d3de3bec74e43eb4b7a94e0ee
Author: Raphael 'kena' Poss <knz@thaumogen.net>
Date:   Tue Mar 3 00:38:55 2020 +0100

    sql: support numeric tuple indexing

    Previously, it was impossible to use `(t).*` to expand a single
    tuple-typed column into multiple projections if the tuple was not
    labeled to start with.

    This limitation was caused by another limitation, that it was not
    possible to refer to a single column in a tuple if the tuple was not
    already labeled.

    This patch addresses both limitations.

    Release note (sql change): CockroachDB now supports expanding all
    columns of a tuple using the `.*` notation, for example `SELECT (t).*
    FROM (SELECT (1,'b',2.3) AS t)`. This is a CockroachDB-specific extension.

    Release note (sql change): CockroachDB now supports accessing the Nth
    column in a column with tuple type using the syntax `(...).@N`, for
    example `SELECT (t).@2 FROM (SELECT (1,'b',2.3) AS t)`. This is a
    CockroachDB-specific extension.

commit 8ffade84564c4d4e071f8a4dcdfc94200a0f612d
Merge: 120d53fc17 79ca3389e1
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 14:08:12 2020 +0000

    Merge #45139

    45139: intervalccl: optimize OverlapCoveringMerge r=dt a=C0rWin

    Change OverlapCoveringMerge instead of iterating over coverings for each
    input range to flatten coverings, sort results and produce ranges with
    one pass. Reducing total complexity from O(nm) to O(n*log(n)).

    Signed-off-by: Artem Barger <bartem@il.ibm.com>

    Release note: none

    Co-authored-by: Artem Barger <bartem@il.ibm.com>

commit 120d53fc17f1f027c6075a2050f2991e369c5293
Merge: 0414b69f64 e2da8bdbbb
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 11:44:30 2020 +0000

    Merge #45587

    45587: log: secondary logger fixes r=ajwerner a=knz

    Needed by / will be exercised by  #45193

    Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>

commit 0414b69f64699bd3537e5f1d4c85e97586fbe227
Merge: 47259c2714 3433fbd782
Author: craig[bot] <bors@cockroachlabs.com>
Date:   Tue Mar 3 09:35:26 2020 +0000

    Merge #45549

    45549: changefeedccl: pick up TargetBytes for backfill ScanRequest r=ajwerner a=ajwerner

    Now that #44925 has merged, we can use a byte limit rather than a row limit.

    Release note: None

    Co-authored-by: Andrew Werner <ajwerner@cockroachlabs.com>

commit 0a1f2516c2021b6e16f4c6b1dbdc41c84f95dc8e
Author: Nathan VanBenschoten <nvanbenschoten@gmail.com>
Date:   Mon Mar 2 16:26:24 2020 -0500

    storage/txnwait: terminate push when pusher aborted at lower epoch

    Closes #40786.
    Closes #44336.

    This commit resolves a bug in distributed deadlock detection that would
    allow a deadlock between transactions to go undetected, stalling the
    workload indefinitely.

    The issue materialized as follows:
    1. two transactions would deadlock and each enter a txnwait queue
    2. they would poll their pushees record along with their own
    3. deadlock detection would eventually pick this up and abort one of the txns
       using the pusher's copy of the txn's proto
    4. however, the aborted txn has since restarted and bumped it epoch
    5. the aborted txn continued to query its record, but failed to ingest any
       updates from it because the record was at a lower epoch than its own
       copy of its txn proto. So it never noticed that it was ABORTED
    6. all other txns in the system including the original contending txn
       piled up behind the aborted txn in the contention queue, waiting for
       it to notice it was aborted and exit the queue
    7. deadlock!

    I'm optimistically closing the two `kv/contention/nodes=4` issues both
    because I hope this is the cause of their recent troubles and also because
    I've been spending a lot of time with the test recently in light of #45482
    and plan to stabilize it fully.

    I plan to backport this to release-19.2. This doesn't need to go all the
    way back to release-19.1 because this was introduces in aed892a.

commit bbac108247a2f723c8ae2bd4c740e4f6bc2fabf5
Author: Jordan Lewis <jordanthelewis@gmail.com>
Date:   Mon Mar 2 23:16:33 2020 -0500

    sql: allow inverted indexes on mixed-case cols

    Previously, a bug prevented creation of inverted indexes on columns with
    mixed-case identifiers. Now, the bug is fixed.

    Release note (bug fix): it is now possible to create inverted indexes on
    columns whose names are mixed-case.

commit cee9355126a55f7c7fc3c80f61e983a164c66dbb
Author: Oliver Tan <otan@cockroachlabs.com>
Date:   Thu Feb 27 10:11:31 2020 -0800

    sql: allow rename database for sequences without a db name reference

    See release note for description. This PR should be included ahead of
    the more "general" fix of changing the DefaultExpr with the new database
    as it unblocks people using
    `experimental_serial_normalization=sql_sequence` from using the database
    rename feature.

    Release note (sql change): Previously, when we renamed a database, any
    table referencing a sequence would be blocked from being able to rename
    the table. This is to block cases where if the table's reference to the
    sequence contains the database name, and the database name changes, we
    have no way of overwriting the table's reference to the sequence in the
    new database. However, if no database name is included in the sequence
    reference, we should continue to allow the database to rename, as is
    implemented with this change.

commit 7383e9dcd342f304376128718f7258129d5af9d7
Author: Radu Berinde <radu@cockroachlabs.com>
Date:   Mon Mar 2 11:15:12 2020 -0800

    opt: extend WithUses and improve NeededMutationCols

    This change extends the WithUses rule prop to keep track of the columns actually
    used by `WithScan`s.

    This set is then used by NeededMutationCols to make sure we never prune mutation
    input columns that are used by FK checks. Currently this never happens, but
    because of fairly subtle reasons: FKs happen to require indexes on both sides,
    and those indexes force the mutation operator to require values for those
    columns. This won't be the case anymore with upsert FK checks, for which this
    fix is required.

    The new information will also be used (in a future change) to prune unused
    columns of `With` itself.

    Release note: None

commit e2da8bdbbb7ecf86df8fe515b82b379139db7844
Author: Raphael 'kena' Poss <knz@thaumogen.net>
Date:   Mon Mar 2 17:04:52 2020 +0100

    log: ensure that the test log scopes works with secondary loggers

    This patch ensures that the file redirection performed by log.Scope /
    log.ScopeWithoutShowLogs is effective with secondary loggers.

    Release note: None

commit 3433fbd7820ac541755c8f6fb8adf9979230f4de
Author: Andrew Werner <ajwerner@cockroachlabs.com>
Date:   Sat Feb 29 13:39:17 2020 -0500

    changefeedccl: pick up TargetBytes for backfill ScanRequest

    Now that #44925 has merged, we can use a byte limit rather than a row limit.

    Release note: None

commit 589cfa101acde9655742617487e4910ed45aac47
Author: Raphael 'kena' Poss <knz@thaumogen.net>
Date:   Mon Mar 2 17:03:18 2020 +0100

    log: ensure secondary loggers get cleaned up

    Previously, if two consecutive tests in a package were using secondary
    loggers, the second test would see the secondary loggers of the first
    test in its logger registry.

    This was problematic because it really made the log files/messages of
    the first test available to the logging file retrieval API.

    This patch cleans this up by de-registering secondary loggers when
    their context is cancelled, which maps to stoppers going down.

    Release note: None

commit 4caee85de6826877b8e55cfb390ea19b7b8a8545
Author: Rohan Yadav <rohany@alumni.cmu.edu>
Date:   Mon Mar 2 11:46:04 2020 -0500

    sql: update error message for primary key change on an interleave parent

    Fixes #45537.

    This PR updates the error message when trying to perform a primary
    key change on an interleaved parent to include the name of the
    table as well as the names of the interleaved children.

    Release note: None

commit 9ac5dbc5b960c5b6dc3b1d4bd522c98fbd4a162f
Author: Céline O'Neil <celineloneil@gmail.com>
Date:   Thu Dec 19 14:02:53 2019 -0500

    opt: factor limit hints into scan and lookup join costs

    Fixes #34811; the example query in this issue now chooses a lookup join
    as desired. The coster now takes limit hints into account when costing
    scans and lookup joins, and propagates limit hints through lookup joins.

    Release note (sql change): The optimizer now considers the likely number
    of rows an operator will need to provide, and might choose query plans
    based on this. In particular, the optimizer might prefer lookup joins
    over alternatives in some situations where all rows of the join will
    probably not be needed.

commit 62c8667569d56ffebf5189e51475eca1d20c9d4a
Author: Raphael 'kena' Poss <knz@thaumogen.net>
Date:   Mon Mar 2 14:35:30 2020 +0100

    cli: reveal SQL error details/hints in CLI commands

    Previously, when a CLI command that uses SQL under the hood
    encountered an error, only the main error message was displayed.

    This patch changes it to reveal the other additional user-visible
    fields.

    Release note (cli change): Certain CLI command now provide more
    details and/or a hint when they encounter an error.

commit f978076cdc2b093e2b6a3205636845ada54e8ff8
Author: Raphael 'kena' Poss <knz@thaumogen.net>
Date:   Wed Dec 11 16:59:31 2019 +0100

    *: allow periods (.) in usernames

    Requested by a user:

    > Currently, there is a restriction for the database username which
    > will limit the certificate-based authentication. It's very common to
    > include .local (e.g.: internal-service2.local) in the CN (Common Name)
    > of a certificate.  The AWS Certificate Manager (ACM) won't even issue
    > a certificate if the "dot" (.) is not present.

    Release note (sql change): Usernames can now contain periods, for
    compatibility with certificate managers that require domain names to
    be used as usernames.

    Release note (security update): The validation rule for principal
    names was extended to support periods and thus allow domainname-like
    principals. For reference, the validation rule is currently the regular
    expression `^[\p{Ll}0-9_][---\p{Ll}0-9_.]+$` and a size limit of 63
    UTF-8 bytes (larger usernames are rejected with an error); for
    comparison, PostgreSQL allows many more characters and truncates at 63
    chacters silently.

commit 79ca3389e15ff11d79caf1f931ccd223cbc7d45b
Author: Artem Barger <bartem@il.ibm.com>
Date:   Sun Feb 16 17:28:58 2020 +0200

    intervalccl: optimize OverlapCoveringMerge

    Change OverlapCoveringMerge instead of iterating over coverings for each
    input range to flatten coverings, sort results and produce ranges with
    one pass. Reducing total complexity from O(nm) to O(n*log(n)).

    Signed-off-by: Artem Barger <bartem@il.ibm.com>

    Release note: none

Release note (<category, see below>): <what> <show> <why>
